### PR TITLE
Margin2

### PR DESCRIPTION
**Pull request by GHesericsu**
Added clients and IDL for Margin Pool, Margin Account, Margin MetaData, Margin Serum, and Margin Swap. 
Added create and load methods for Margin Pool and Margin Account

### CODE REVIEW

**Comment by jetjake** [comment]
You have a conflict ^

File: src/margin/client.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { Program, Provider } from "@project-serum/anchor"
+import { PublicKey } from "@solana/web3.js"
+import { connect, Hooks } from "../common"
+import { JetMarginIdl } from "./idl"
+
+export class MarginClient {
+  static readonly MARGIN_PROGRAM_ID = new PublicKey("JPMRGNgRk3w2pzBM1RLNBnpGxQYsFQ3yXKpuk4tTXVZ")
```
Comment (jetjake) [inline src/margin/client.ts:7]: Is the PROGRAM_ID the same value no matter the environment (local, devnet, mainnet, etc.)?

File: src/marginSerum/idl.ts [hunk 2]
```diff
@@ -0,0 +1,397 @@
+export type JetMarginSerumIdl = {
```
Comment (jetjake) [inline src/marginSerum/idl.ts:1]: Is this file autogenerated? If so what is the autogeneration based on?

File: src/staking/client.ts [hunk 3]
```diff
@@ -44,7 +45,7 @@ export class StakeClient {
    * @returns {(Program<Idl> | undefined)}
```
Comment (jetjake) [inline src/staking/client.ts:45]: Does this need to be updated as well for posterity

File: src/staking/client.ts [hunk 4]
```diff
@@ -44,7 +45,7 @@ export class StakeClient {
...
    * @memberof StakeClient
    */
-  static use(provider: Provider): Program<Idl> | undefined {
+  static use(provider: Provider): Program<StakeIdl> | undefined {
```
Comment (jetjake) [inline src/staking/client.ts:48]: I guess I wonder, before this change, how did it function properly, or did it?

File: src/marginPool/client.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Program, Provider } from "@project-serum/anchor"
+import { PublicKey } from "@solana/web3.js"
+import { connect, Hooks } from "../common"
+import { JetMarginPoolIdl } from "./idl"
+
+export class MarginPoolClient {
```
Comment (jetjake) [inline src/marginPool/client.ts:6]: Whats the difference between MarginPoolClient and MarginPool class?  MarginPoolClient loads the MarginPool program for which then can be interacted with (create, load, deriveAccounts)?

**Review by jetjake — COMMENTED** [review]
Code looks nice and tight.   All of my questions are for my own knowledge and are not questioning the functionality.

File: src/margin/client.ts (hunk 1 above)
Comment (GHesericsu) [reply src/margin/client.ts:7]: there will be different programIds for each net

File: src/marginSerum/idl.ts (hunk 2 above)
Comment (GHesericsu) [reply src/marginSerum/idl.ts:1]: yes this file is generated by running `anchor build` on jet-margin repo. You can find the files under target/idl https://book.anchor-lang.com/chapter_4/cli.html?highlight=target#build

File: src/staking/client.ts (hunk 3 above)
Comment (GHesericsu) [reply src/staking/client.ts:45]: can you elaborate?

File: src/staking/client.ts (hunk 4 above)
Comment (GHesericsu) [reply src/staking/client.ts:48]: I'm not sure.... I just saw an error and fixed it. The original changes were made by someone else.

File: src/marginPool/client.ts (hunk 5 above)
Comment (GHesericsu) [reply src/marginPool/client.ts:6]: Ya it loads the program with anchor IDL.

File: src/margin/client.ts (hunk 1 above)
Comment (GHesericsu) [reply src/margin/client.ts:7]: i just don't think we have devnet programs deployed yet. and localnet is not that useful for the front end

Comment (jetjake) [reply src/margin/client.ts:7]: I was wondering if we move that PublicKey to a configuration file which could be loaded depending on environment

File: src/staking/client.ts (hunk 3 above)
Comment (jetjake) [reply src/staking/client.ts:45]: Just talking about updating the comment that specifies the return value matches the actual return value
`@returns {(Program<Idl> | undefined)}`
to 
`@returns {(Program<StakeIdl> | undefined)}`

File: src/margin/client.ts (hunk 1 above)
Comment (GHesericsu) [reply src/margin/client.ts:7]: ya I think we will. this is just temporary.

File: src/staking/client.ts (hunk 3 above)
Comment (GHesericsu) [reply src/staking/client.ts]: fixed

File: src/margin/client.ts (hunk 1 above)
Comment (Tristyn) [reply src/margin/client.ts:7]: Our program id's don't change, as well as the token program and most projects. Serum does change which makes it annoying

File: src/staking/client.ts (hunk 4 above)
Comment (Tristyn) [reply src/staking/client.ts:48]: Program<StakeIdl> doesn't change any behaviour, but gives typescript stricter type checking based on the type in idl.ts

File: src/margin/marginAccount.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import { Program } from "@project-serum/anchor"
+import { findDerivedAccount } from "../common"
+import { Hooks } from "../common/hooks"
+import { MarginAccountInfo } from "./state"
+import { checkNull } from "../common/index"
+
+export class MarginAccount {
+  constructor(public marginProgram: Program, public address: PublicKey, public marginAccount: MarginAccountInfo) {}
+
+  //load multiple accounts?`
```
Comment (Tristyn) [inline src/margin/marginAccount.ts:11]:
```suggestion
```

File: src/common/associatedToken.ts [hunk 7]
```diff
@@ -7,6 +7,7 @@ import { parseMintAccount, parseTokenAccount } from "./accountParser"
 import { Hooks } from "./hooks"
 import { findDerivedAccount } from "."
 
+//ADD wrapping and unwrapping SOL
```
Comment (Tristyn) [inline src/common/associatedToken.ts:10]:
```suggestion
```

File: src/marginPool/marginPool.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    return Hooks.usePromise(async () => program && MarginPool.load(program, tokenMint), [program])
+  }
+
+  /**
+   * Derive accounts from tokenMint
+   * @param {PublicKey} programId
+   * @param {PublicKey} tokenMint
+   * @returns {PublicKey} Margin Pool Address
+   */
+  private static deriveAccounts(programId: PublicKey, tokenMint: PublicKey): MarginPoolAddresses {
```
Comment (Tristyn) [inline src/marginPool/marginPool.ts:72]:
```suggestion
  private static derive(programId: PublicKey, tokenMint: PublicKey): MarginPoolAddresses {
```

File: src/marginPool/marginPool.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  ) {}
+
+  /**
+   * Load a Margin Pool Program Account
+   * @param {Program} program
+   * @param {PublicKey} tokenMint
+   * @returns {Promise<MarginPool>}
+   */
+  static async load(program: Program, tokenMint: PublicKey): Promise<MarginPool> {
+    const addresses = this.deriveAccounts(program.programId, tokenMint)
```
Comment (Tristyn) [inline src/marginPool/marginPool.ts:37]:
```suggestion
    const addresses = this.derive(program.programId, tokenMint)
```

File: src/marginPool/marginPool.ts [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  static create(
+    program: Program,
+    tokenMetaDataInfo: TokenMetadataInfo,
+    authority: PublicKey,
+    params: CreatePoolParams,
+    feeDestination: PublicKey,
+    marginPoolConfig: MarginPoolConfig
+  ): Promise<string> {
+    //derive pool accounts
+    const addresses = this.deriveAccounts(program.programId, tokenMetaDataInfo.tokenMint)
```
Comment (Tristyn) [inline src/marginPool/marginPool.ts:107]:
```suggestion
    const addresses = this.derive(program.programId, tokenMetaDataInfo.tokenMint)
```

File: src/staking/unbondingAccount.ts [hunk 11]
```diff
@@ -7,7 +7,7 @@ import { Hooks } from "../common/hooks"
 import { TOKEN_PROGRAM_ID } from "@solana/spl-token"
 
 export interface UnbondingAccountInfo {
-  /// The related account requesting to unstake
+  /// The related account requesting to un
```
Comment (Tristyn) [inline src/staking/unbondingAccount.ts:10]:
```suggestion
  /// The related account requesting to unstake
```

**Review by Tristyn — APPROVED** [review]
Well done, you can move some of the comments out of the code (But I agree wrapping and unwrapping sol would be great to have) but this is well put together ⭐

**Pull request merged**